### PR TITLE
315-ask-answer-type-first-when-creating-question-in-admin

### PR DESCRIPTION
**Pull request by aliuk2012**
#### What problem does the pull request solve?

Trello card: https://trello.com/c/N78PUM5K/315-ask-answer-type-first-when-creating-question-in-admin

This PR introduces a new flow for creating a question for a form. This includes a page where the user selects the answer type for their new question, and then sets other question details. To do this, the answer type is stored in session, which is new for this application. 

**Add a question**

To test, add a new question to a form:

image

You should then be able to select an answer type:

image

And on submission, you should be able to review the answer type that's been selected:

image

**Edit an existing question**

Click on edit for an existing question:

image

You are able to see the answer type in the summary, with a change link:

image

Clicking on the change link will bring you to the pre-populated answer type list:

image

Selecting a new answer type will save this to the API, and return to the Edit page.

#### Checklist

- [ ] I've used the pull request template
- [ ] I've linked this PR to the relevant issue (if mission work)
- [ ] I've written unit tests for these changes (if code change)
- [ ] I've updated the documentation in (If any documentation requires updating)
  - [ ] README.md
  - [ ] Elsewhere (please link)

### CODE REVIEW

File: app/controllers/pages_controller.rb [hunk 1]
```diff
@@ -19,12 +20,15 @@ def create
 
   def edit
     @page = Page.find(params[:page_id], params: { form_id: @form.id })
+
+    # if session[:page]["id"] == @page.id
+    #   session[:page]["answer_type"]
```
Comment (DavidBiddle) [inline app/controllers/pages_controller.rb:25]: Can these comments be removed?

Comment (SamJamCul) [reply app/controllers/pages_controller.rb]: Gone!

File: app/views/pages/_form.html.erb [hunk 2]
```diff
@@ -1,27 +1,38 @@
...
+
+  <% end %>
+
+  <%= govuk_summary_list do |summary_list|
+    summary_list.row do |row|
+      row.key(text: "Answer type")
+      row.value(text: t("helpers.label.page.answer_type_options.names.#{page_object.answer_type}"))
+      row.action(text: "Change",
+                 href: change_answer_type_path,
+                 visually_hidden_text: t("helpers.label.page.answer_type_options.names.#{page_object.answer_type}")
```
Comment (DavidBiddle) [inline app/views/pages/_form.html.erb:28]: I think  we want the hidden text to include 'answer type' here - that's how it is in the prototype. Think the logic here is that someone using speech recognition is probably more likely to say 'Click change answer type text' or 'Click change answer type' than 'Click change text'.

```suggestion
                 visually_hidden_text: "Answer type" + t("helpers.label.page.answer_type_options.names.#{page_object.answer_type}")
```

(there's probably a nicer way of concatenating these strings than that)

**Review by DavidBiddle — COMMENTED** [review]
Looks good overall, have just raised an issue with some visually hidden text

File: app/controllers/pages_controller.rb [hunk 3]
```diff
@@ -35,8 +36,8 @@ def update
 
 private
 
-  def page_params(form_id)
-    params.require(:page).permit(:question_text, :question_short_name, :hint_text, :answer_type, :is_optional).merge(form_id:)
+  def page_params
+    params.require(:page).permit(:question_text, :question_short_name, :hint_text, :answer_type, :is_optional).merge(form_id: @form.id).merge(session:)
```
Comment (thomasiles) [inline app/controllers/pages_controller.rb:40]: Do we need to merge the whole session here or just add `session[:answer_type]`? I think the session has other stuff in it - e.g. csrf token etc.

File: app/views/pages/_form.html.erb (hunk 2 above)
Comment (SamJamCul) [reply app/views/pages/_form.html.erb:28]: Good spot - after some messing around I couldn't find a better way of interpolating within an interpolation, so the `+` seems like the easiest way of doing it.

**Review by DavidBiddle — APPROVED** [review]

File: app/forms/forms/type_of_answer_form.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    session[:page] = { answer_type: }
+  end
+
+  def number(form)
+    # If this page is in form, return the position, else it must be new so
+    # return the number if it was inserted at the end
+    index = form.pages.index(self)
+    (index.nil? ? form.pages.length : index) + 1
+  end
```
Comment (thomasiles) [inline app/forms/forms/type_of_answer_form.rb:20]: `number(form)` is now in two places. - /models/Page.rb and here. Could we put it somewhere better? would it make more sense in /models/Form.rb and taking page as an argument?

**Review by thomasiles — APPROVED** [review]
Impressive work!

File: app/forms/forms/type_of_answer_form.rb (hunk 4 above)
Comment (SamJamCul) [reply app/forms/forms/type_of_answer_form.rb]: This has now been refactored into `/models/Form.rb`, as well as flushing out a bug with the page number caption for editing the answer type!

**Pull request merged**